// @autonomys/auto-drive/dist/api/models/uploads.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.folderUploadSchema = exports.EncryptionAlgorithm = exports.CompressionAlgorithm = exports.fileUploadSchema = exports.UploadStatus = exports.UploadType = void 0;
const zod_1 = require("zod");
const folderTree_1 = require("./folderTree");
var UploadType;
(function (UploadType) {
    UploadType["FILE"] = "file";
    UploadType["FOLDER"] = "folder";
})(UploadType || (exports.UploadType = UploadType = {}));
var UploadStatus;
(function (UploadStatus) {
    UploadStatus["PENDING"] = "pending";
    UploadStatus["COMPLETED"] = "completed";
    UploadStatus["MIGRATING"] = "migrating";
    UploadStatus["CANCELLED"] = "cancelled";
    UploadStatus["FAILED"] = "failed";
})(UploadStatus || (exports.UploadStatus = UploadStatus = {}));
exports.fileUploadSchema = zod_1.z.object({
    id: zod_1.z.string(),
    rootId: zod_1.z.string(),
    relativeId: zod_1.z.string().nullable(),
    type: zod_1.z.nativeEnum(UploadType),
    status: zod_1.z.nativeEnum(UploadStatus),
    fileTree: zod_1.z.null(),
    name: zod_1.z.string(),
    mimeType: zod_1.z.string().nullable(),
    oauthProvider: zod_1.z.string(),
    oauthUserId: zod_1.z.string(),
});
var CompressionAlgorithm;
(function (CompressionAlgorithm) {
    CompressionAlgorithm["ZLIB"] = "ZLIB";
})(CompressionAlgorithm || (exports.CompressionAlgorithm = CompressionAlgorithm = {}));
var EncryptionAlgorithm;
(function (EncryptionAlgorithm) {
    EncryptionAlgorithm["AES_256_GCM"] = "AES_256_GCM";
})(EncryptionAlgorithm || (exports.EncryptionAlgorithm = EncryptionAlgorithm = {}));
exports.folderUploadSchema = zod_1.z.object({
    id: zod_1.z.string(),
    rootId: zod_1.z.string(),
    relativeId: zod_1.z.string().nullable(),
    type: zod_1.z.nativeEnum(UploadType),
    status: zod_1.z.nativeEnum(UploadStatus),
    fileTree: folderTree_1.FolderTreeFolderSchema,
    name: zod_1.z.string(),
    mimeType: zod_1.z.null(),
    oauthProvider: zod_1.z.string(),
    oauthUserId: zod_1.z.string(),
    uploadOptions: zod_1.z.null(),
});
