// @autonomys/auto-drive/dist/utils/async.js
"use strict";
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __await = (this && this.__await) || function (v) { return this instanceof __await ? (this.v = v, this) : new __await(v); }
var __asyncGenerator = (this && this.__asyncGenerator) || function (thisArg, _arguments, generator) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var g = generator.apply(thisArg, _arguments || []), i, q = [];
    return i = Object.create((typeof AsyncIterator === "function" ? AsyncIterator : Object).prototype), verb("next"), verb("throw"), verb("return", awaitReturn), i[Symbol.asyncIterator] = function () { return this; }, i;
    function awaitReturn(f) { return function (v) { return Promise.resolve(v).then(f, reject); }; }
    function verb(n, f) { if (g[n]) { i[n] = function (v) { return new Promise(function (a, b) { q.push([n, v, a, b]) > 1 || resume(n, v); }); }; if (f) i[n] = f(i[n]); } }
    function resume(n, v) { try { step(g[n](v)); } catch (e) { settle(q[0][3], e); } }
    function step(r) { r.value instanceof __await ? Promise.resolve(r.value.v).then(fulfill, reject) : settle(q[0][2], r); }
    function fulfill(value) { resume("next", value); }
    function reject(value) { resume("throw", value); }
    function settle(f, v) { if (f(v), q.shift(), q.length) resume(q[0][0], q[0][1]); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bufferToIterable = exports.fileToIterable = exports.asyncFromStream = exports.asyncByChunk = void 0;
const buffer_1 = require("buffer");
const asyncByChunk = function (asyncIterable, chunkSize) {
    return __asyncGenerator(this, arguments, function* () {
        var _a, e_1, _b, _c;
        let buffer = buffer_1.Buffer.alloc(0);
        try {
            for (var _d = true, asyncIterable_1 = __asyncValues(asyncIterable), asyncIterable_1_1; asyncIterable_1_1 = yield __await(asyncIterable_1.next()), _a = asyncIterable_1_1.done, !_a; _d = true) {
                _c = asyncIterable_1_1.value;
                _d = false;
                const chunk = _c;
                buffer = buffer_1.Buffer.concat([buffer, chunk]);
                while (buffer.length >= chunkSize) {
                    yield yield __await(buffer.subarray(0, chunkSize));
                    buffer = buffer.subarray(chunkSize);
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (!_d && !_a && (_b = asyncIterable_1.return)) yield __await(_b.call(asyncIterable_1));
            }
            finally { if (e_1) throw e_1.error; }
        }
        if (buffer.length > 0) {
            yield yield __await(buffer);
        }
    });
};
exports.asyncByChunk = asyncByChunk;
const asyncFromStream = function (stream) {
    return __asyncGenerator(this, arguments, function* () {
        const reader = stream.getReader();
        let result = yield __await(reader.read());
        while (!result.done) {
            yield yield __await(buffer_1.Buffer.from(result.value));
            result = yield __await(reader.read());
        }
    });
};
exports.asyncFromStream = asyncFromStream;
const fileToIterable = function (file) {
    return __asyncGenerator(this, arguments, function* () {
        const chunkSize = 1024 * 1024;
        for (let i = 0; i < file.size; i += chunkSize) {
            yield yield __await(buffer_1.Buffer.from(yield __await(file.slice(i, i + chunkSize).arrayBuffer())));
        }
    });
};
exports.fileToIterable = fileToIterable;
const bufferToIterable = function (buffer) {
    return __asyncGenerator(this, arguments, function* () {
        yield yield __await(buffer);
    });
};
exports.bufferToIterable = bufferToIterable;
